### DISCUSSION TITLE
Volume Control Does not Seem to Work

### DISCUSSION THREAD

**Discussion by RickLloyd24** [pschatzmann/arduino-audio-tools#1593] [Q&A] [answered]
Sorry if this is a stupid question.  I am using the file streams-i2s-filter-i2s.ino example.  Everything works but the volume is very low.  I tried adding a VolumeStream and used multiple values (0.3 to 0.9)  but it did not change the volume.  What am I doing wrong?  I enclosed the complete code.

On other questions what is the maximum number of coefficients you can use?

#include "AudioTools.h"
AudioInfo info(32000, 2, 16);
I2SStream in;
I2SStream out; 
VolumeStream gain(in);

// copy filtered values
FilteredStream<int16_t, float> filtered(in, info.channels);  // Defiles the filter as BaseConverter
StreamCopy copier(out, filtered); // copies sound into i2s

// define FIR filter parameters
float coef[] = { 0.021, 0.096, 0.146, 0.096, 0.021};

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Info); 

  // setup filters for all available channels
  filtered.setFilter(0, new FIR<float>(coef));
  filtered.setFilter(1, new FIR<float>(coef));

  // start I2S in
  Serial.println("starting I2S...");
  auto config_in = in.defaultConfig(RX_MODE);
  config_in.copyFrom(info); 
  config_in.i2s_format = I2S_STD_FORMAT;
  config_in.is_master = true;
  config_in.port_no = 0;
  config_in.pin_ws = 22;
  config_in.pin_bck = 23;
  config_in.pin_data = 21;
  // config_in.fixed_mclk = sample_rate * 256
  // config_in.pin_mck = 3

  in.begin(config_in);

  // start I2S out
  auto config_out = out.defaultConfig(TX_MODE);
  config_out.copyFrom(info); 
  config_out.i2s_format = I2S_STD_FORMAT;
  config_out.is_master = true;
  config_out.port_no = 1;
  config_out.pin_ws = 25;
  config_out.pin_bck = 14;
  config_out.pin_data = 26;
  out.begin(config_out);

  // set initial volume
  gain.begin(info); // we need to provide the bits_per_sample and channels
  gain.setVolume(0.9);

  Serial.println("I2S started...");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}

**Comment by pschatzmann** [comment]
gain is not part of the processing chain, so it does not have any impact!

**Reply by RickLloyd24** [reply, marked as answer]
How do I make gain part of the processing chain?

**Reply by RickLloyd24** [reply]
I understand your response and was able to get it working.
